### PR TITLE
conversion to use tf2 public API - installment 1

### PR DESCRIPTION
**Pull request by armando-fandango**
- as per #134, replaces calls with TF2 public API
- TODO: replace v1 / compat.v1 constructs with v2
- TODO: `my_decoder is not callable` produced by pylint for decoder_test.py:63,140
- dense_image_warp_tests are failing - fixing them was not part of this pull request

### CODE REVIEW

**Comment by armando-fandango** [comment]
If I revert back to tensorflow.python.keras.layer then python 2 lint errors disappear, but python 3 still compalins about being non-callable.

**Comment by seanpmorgan** [comment]
> If I revert back to tensorflow.python.keras.layer then python 2 lint errors disappear, but python 3 lint still complains about being non-callable.

Its preferable to ignore pylint errors than to utilize private APIs. See https://github.com/tensorflow/addons/pull/137

**Comment by armando-fandango** [comment]
> Its preferable to ignore pylint errors than to utilize private APIs. See #137

Added pylint disable for not-callable in test file. This pull is ready to be reviewed. Not everything is converted, hence the name installment-1. Once this is merged then I will work on converting the remaining ones too.

File: tensorflow_addons/seq2seq/attention_wrapper.py [hunk 1]
```diff
@@ -288,8 +277,8 @@ def _setup_memory(self,
             raise ValueError(
                 "memory_sequence_length and memory_mask cannot be "
                 "used at same time for attention.")
-        with ops.name_scope(self.name, "BaseAttentionMechanismInit",
-                            nest.flatten(memory)):
+        with tf.name_scope(self.name or "BaseAttentionMechanismInit"):
+            #tf.nest.flatten(memory)):
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/attention_wrapper.py]: This can be removed I think.

File: tensorflow_addons/seq2seq/attention_wrapper.py [hunk 2]
```diff
@@ -831,13 +819,12 @@ def safe_cumprod(x, *args, **kwargs):
     Returns:
       Cumulative product of x.
     """
-    with ops.name_scope(None, "SafeCumprod", [x]):
-        x = ops.convert_to_tensor(x, name="x")
+    with tf.name_scope("SafeCumprod"):  #, [x]):
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/attention_wrapper.py]: same.

File: tensorflow_addons/seq2seq/attention_wrapper.py [hunk 3]
```diff
@@ -1499,14 +1483,13 @@ def hardmax(logits, name=None):
     Returns:
       A batched one-hot tensor.
     """
-    with ops.name_scope(name, "Hardmax", [logits]):
-        logits = ops.convert_to_tensor(logits, name="logits")
-        if tensor_shape.dimension_value(logits.get_shape()[-1]) is not None:
-            depth = tensor_shape.dimension_value(logits.get_shape()[-1])
+    with tf.name_scope(name or "Hardmax"):  #, [logits]):
+        logits = tf.convert_to_tensor(logits, name="logits")
+        if tf.compat.dimension_value(logits.get_shape()[-1]) is not None:
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/attention_wrapper.py]: did u miss v1 here?

File: tensorflow_addons/seq2seq/beam_search_decoder.py [hunk 4]
```diff
@@ -226,16 +214,16 @@ def _check_batch_beam(t, batch_size, beam_width):
...
-        condition = math_ops.logical_or(
-            math_ops.equal(shape[1], batch_size * beam_width),
-            math_ops.logical_and(
-                math_ops.equal(shape[1], batch_size),
-                math_ops.equal(shape[2], beam_width)))
-    return control_flow_ops.Assert(condition, [error_message])
+        condition = tf.logical_or(
+            tf.equal(shape[1], batch_size * beam_width),
+            tf.logical_and(
+                tf.equal(shape[1], batch_size), tf.equal(shape[2],
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/beam_search_decoder.py]: probably wrap the second tf.equal into next line to make it more readable.

File: tensorflow_addons/seq2seq/attention_wrapper.py (hunk 1 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/attention_wrapper.py]: Done.

File: tensorflow_addons/seq2seq/attention_wrapper.py (hunk 2 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/attention_wrapper.py]: done

File: tensorflow_addons/seq2seq/attention_wrapper.py (hunk 3 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/attention_wrapper.py]: v1 and v2 are aliases for the same `tf.compat.dimension_value`

File: tensorflow_addons/seq2seq/beam_search_decoder.py (hunk 4 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/beam_search_decoder.py]: done

File: tensorflow_addons/seq2seq/attention_wrapper.py [hunk 5]
```diff
@@ -1832,10 +1812,9 @@ def get_initial_state(self, inputs=None, batch_size=None, dtype=None):
             to the wrapper object at initialization time.
         """
         if inputs is not None:
-            batch_size = array_ops.shape(inputs)[0]
+            batch_size = tf.shape(inputs)[0]
             dtype = inputs.dtype
-        with ops.name_scope(
-                type(self).__name__ + "ZeroState", values=[batch_size]):  # pylint: disable=bad-continuation
+        with tf.name_scope(type(self).__name__ + "ZeroState"):  #, values=[batch_size]):  # pylint: disable=bad-continuation
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/attention_wrapper.py]: Please cleanup here as well.

File: tensorflow_addons/seq2seq/beam_search_decoder.py [hunk 6]
```diff
@@ -120,9 +113,9 @@ def tile_batch(t, multiplier, name=None):
       ValueError: if tensor(s) `t` do not have a statically known rank or
       the rank is < 1.
     """
-    flat_t = nest.flatten(t)
-    with ops.name_scope(name, "tile_batch", flat_t + [multiplier]):
-        return nest.map_structure(lambda t_: _tile_batch(t_, multiplier), t)
+    flat_t = tf.nest.flatten(t)
+    with tf.name_scope(name or "tile_batch"):  #, flat_t + [multiplier]):
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/beam_search_decoder.py]: same here.

File: tensorflow_addons/seq2seq/beam_search_decoder.py [hunk 7]
```diff
@@ -563,21 +548,22 @@ def step(self, time, inputs, state, name=None):
         length_penalty_weight = self._length_penalty_weight
         coverage_penalty_weight = self._coverage_penalty_weight
 
-        with ops.name_scope(name, "BeamSearchDecoderStep",
-                            (time, inputs, state)):
+        with tf.name_scope(name or "BeamSearchDecoderStep"):
+            #,(time, inputs, state)):
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/beam_search_decoder.py]: and here.

File: tensorflow_addons/seq2seq/loss.py [hunk 8]
```diff
@@ -107,45 +103,48 @@ def sequence_loss(logits,
         raise ValueError(
             "average_across_batch and sum_over_batch cannot be set "
             "to True at same time.")
-    with ops.name_scope(name, "sequence_loss", [logits, targets, weights]):
-        num_classes = array_ops.shape(logits)[2]
-        logits_flat = array_ops.reshape(logits, [-1, num_classes])
-        targets = array_ops.reshape(targets, [-1])
+    with tf.name_scope(name
+                       or "sequence_loss"):  #, [logits, targets, weights]):
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/loss.py]: and here.

File: tensorflow_addons/seq2seq/sampler.py [hunk 9]
```diff
@@ -712,71 +692,68 @@ def next_inputs(self, time, outputs, state, sample_ids):
 # categorical_sample) mimic TensorFlow Probability distribution semantics.
 def _call_sampler(sample_n_fn, sample_shape, name=None):
     """Reshapes vector of samples."""
-    with ops.name_scope(name, "call_sampler", values=[sample_shape]):
-        sample_shape = ops.convert_to_tensor(
-            sample_shape, dtype=dtypes.int32, name="sample_shape")
+    with tf.name_scope(name or "call_sampler"):  #, values=[sample_shape]):
```
Comment (qlzh727) [inline tensorflow_addons/seq2seq/sampler.py]: finally here.

File: tensorflow_addons/seq2seq/attention_wrapper.py (hunk 5 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/attention_wrapper.py]: done

File: tensorflow_addons/seq2seq/beam_search_decoder.py (hunk 7 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/beam_search_decoder.py]: done

File: tensorflow_addons/seq2seq/loss.py (hunk 8 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/loss.py]: done

File: tensorflow_addons/seq2seq/sampler.py (hunk 9 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/sampler.py]: done

File: tensorflow_addons/seq2seq/beam_search_decoder.py (hunk 6 above)
Comment (armando-fandango) [reply tensorflow_addons/seq2seq/beam_search_decoder.py]: done

**Comment by qlzh727** [comment]
Thanks for the update, Looks good to me now. Not sure why the tests are failing, seems that the error is from custom c ops. @seanpmorgan

**Comment by seanpmorgan** [comment]
Tests are failing because of this: https://github.com/tensorflow/tensorflow/issues/27697

Not much we can do other than wait for a fix. That or pin to a nightly version from a day or two ago as horovod has done.

**Pull request merged**